core/src/main/scala-3/spire/syntax/macros/cforMacros.scala

```diff
@@ -15,157 +15,117 @@
-    var index = $fromExpr
-    val limit = $untilExpr
-    while index > limit do
-      ${ Expr.betaReduce(body) }(index)
-      index -= $stride
-  }
+  def zeroStride[T, R](orig: Expr[R])(using Quotes): Code[T] = _ =>
+    import quotes.reflect.*
+    report.error("zero stride", orig)
+    '{}
 
-  r match
-    case '{ ($i: Long) until $j } => strideUpUntil(i, j, Expr(1L))
-    case '{ ($i: Long) to $j }    => strideUpTo(i, j, Expr(1L))
-    case '{ ($i: Long) until $j by $step } =>
-      step.asTerm match {
-        case Literal(LongConstant(k)) if k > 0  => strideUpUntil(i, j, Expr(k))
-        case Literal(LongConstant(k)) if k < 0  => strideDownUntil(i, j, Expr(-k))
-        case Literal(LongConstant(k)) if k == 0 => report.error("zero stride", step); '{}
-        case _ =>
-          report.warning(s"defaulting to foreach, can not optimise non-constant step", step)
-          '{ val b = $body; $r.foreach(b) }
-      }
-    case '{ ($i: Long) to $j by $step } =>
-      step.asTerm match {
-        case Literal(LongConstant(k)) if k > 0  => strideUpTo(i, j, Expr(k))
-        case Literal(LongConstant(k)) if k < 0  => strideDownTo(i, j, Expr(-k))
-        case Literal(LongConstant(k)) if k == 0 => report.error("zero stride", step); '{}
-        case _ =>
-          report.warning(s"defaulting to foreach, can not optimise non-constant step", step)
-          '{ val b = $body; $r.foreach(b) }
-      }
+  def deOpt[T, R](orig: Expr[R], foreach: Expr[Unit])(using Quotes): Code[T] = _ =>
+    import quotes.reflect.*
+    report.warning(s"defaulting to foreach, can not optimise range expression", orig)
+    foreach
 
-    case _ =>
-      report.warning(s"defaulting to foreach, can not optimise range expression", r)
-      '{ val b = $body; $r.foreach(b) }
+  trait CanLoop[T]:
+    extension (x: Expr[T]) def stepBy(y: Expr[T])(using Quotes): Expr[T]
 
-end fastForRangeMacroLong
+  object CanLoop:
+    given CanLoop[Int] with
+      extension (x: Expr[Int]) def stepBy(y: Expr[Int])(using Quotes): Expr[Int] = '{ $x + $y }
 
-def fastForRangeMacro(r: Expr[Range], body: Expr[Int => Unit])(using quotes: Quotes): Expr[Unit] =
-  import quotes._
-  import quotes.reflect._
+    given CanLoop[Long] with
+      extension (x: Expr[Long]) def stepBy(y: Expr[Long])(using Quotes): Expr[Long] = '{ $x + $y }
 
-  def strideUpUntil(fromExpr: Expr[Int], untilExpr: Expr[Int], stride: Expr[Int]): Expr[Unit] = '{
-    var index = $fromExpr
-    val limit = $untilExpr
-    while (index < limit) {
-      ${ Expr.betaReduce(body) }(index)
-      index += $stride
-    }
-  }
+end RangeForImpl
 
-  def strideUpTo(fromExpr: Expr[Int], untilExpr: Expr[Int], stride: Expr[Int]): Expr[Unit] = '{
-    var index = $fromExpr
-    val end = $untilExpr
-    while index <= end do
-      ${ Expr.betaReduce(body) }(index)
-      index += $stride
-  }
+private def letFunc[A, B, C](using Quotes)(name: String, rhs: Expr[A => B])(in: Expr[A => B] => Expr[C]): Expr[C] =
```

Comment:
Would you mind writing a comment(s) to explain a little of the macro magic here? For curious users and future contributors/maintainers alike :)

Reply:
I have added some doc comments, and also found an extractor I can use to simplify function literal detection